xarray/core/weighted.py

```diff
@@ -239,6 +290,87 @@ def _weighted_std(
 
         return cast("DataArray", np.sqrt(self._weighted_var(da, dim, skipna)))
 
+    def _weighted_quantile(
+        self,
+        da: "DataArray",
+        q,
+        dim: Optional[Union[Hashable, Iterable[Hashable]]] = None,
+        skipna: Optional[bool] = None,
+        keep_attrs: Optional[bool] = None,
+    ) -> "DataArray":
+        """Apply a weighted ``quantile`` to a DataArray along some dimension(s)."""
+
+        def _weighted_quantile_type7_1d(data, weights, q, skipna):
+            # This algorithm has been adapted from:
+            #   https://aakinshin.net/posts/weighted-quantiles/#reference-implementation
+            not_nan = ~np.isnan(data)
+            if skipna:
+                # Remove nans from data and weights
+                data = data[not_nan]
+                weights = weights[not_nan]
+            elif ~not_nan.any():
+                # Return nan if data contains any nan
+                return np.full(q.size, np.nan)
+            # Filter out data (and weights) associated with zero weights, which also flattens them
+            nonzero_weights = weights != 0
+            data = data[nonzero_weights]
+            weights = weights[nonzero_weights]
+            n = data.size
+            if n == 0:
+                # Possibly empty after nan or zero weight filtering above
+                return np.full(q.size, np.nan)
+            weights = weights / weights.sum()
+            sorter = np.argsort(data)
+            data = data[sorter]
+            weights = weights[sorter]
+            weights_cum = np.append(0, weights.cumsum())
+            q = np.atleast_2d(q).T
+            h = q * (n - 1) + 1
+            u = np.maximum((h - 1) / n, np.minimum(h / n, weights_cum))
+            v = u * n - h + 1
+            w = np.diff(v)
+            r = (data * w).sum(axis=1)
+
+            return r
+
+        if da.shape != self.weights.shape:
+            raise ValueError("da and weights must have the same shape")
```

Comment:
I think `da` and `self.weights` must be broadcast-able and not necessarily have the same shape. Potentially `apply_ufunc` takes care of this? (But I am not sure and would need to test).

Reply:
I had to explicitly call align to make sure coordinates for da and weights match.

Reply:
I agree with the principle, but in practice I'm having trouble getting it to work. Your first example doesn't work out of the box. I get 
```python
ValueError: operand to apply_ufunc has required core dimensions ['dim_0', 'dim_1'], but some of these dimensions are absent on an input variable: ['dim_1']
```